### PR TITLE
Implement UE distribution inside polygons

### PR DESCRIPTION
**Pull request by CalilQ**
Topography grid was also implemented, to define stations’ z coordinate. Polygons are parsed from kml file and used to create a mask on top of the topography grid. UEs are distributed inside the mask.

### CODE REVIEW

File: sharc/map/topography.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        lin_f = (y - self.low_left[1]) / self.resolution
+        col_f = (x - self.low_left[0]) / self.resolution
+        lin = self.nrows - lin_f.astype(int) - 1
+        col = col_f.astype(int)
+        
+        return self.topography_grid[lin,col]
+
+
+if __name__ == '__main__':
+    file = '../parameters/maps/Brucutu_res_20m.asc'
```
Comment (brunohcfaria) [inline sharc/map/topography.py]: Need to fix this hardcoded path.

File: sharc/map/topography.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    print(topo_data)
+
+    print(topo_data.topography_grid)
+
+    # Customize the color map so the invalid value '13' is grey
+    custom_map = plt.cm.get_cmap('Blues', 7)
+    cmaplist = [custom_map(i) for i in range(custom_map.N)]
+    # reverse the colors
+    cmaplist = cmaplist[::-1]
+    # add grey to the last color
```
Comment (brunohcfaria) [inline sharc/map/topography.py]: Remove those comments

File: sharc/map/topography.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+    # The invalid value is -9999 but this specific data set has invalid values that are below 300
+    # masked_grid = np.ma.masked_where(topo_grid == -9999, topo_grid)
+    masked_grid = np.ma.masked_where(topo_grid <= 400, topo_grid)
+
+    plt.imshow(masked_grid, cmap=custom_map, extent=[topo_data.low_left[0],
+                                                  topo_data.low_left[0] + topo_data.ncols*topo_data.resolution,
+                                                  topo_data.low_left[1],
+                                                  topo_data.low_left[1] + topo_data.nrows*topo_data.resolution])
+    # plt.imshow(masked_grid, cmap=custom_map)
```
Comment (brunohcfaria) [inline sharc/map/topography.py]: Remove this commented line

File: sharc/topology/topology_input_map.py [hunk 4]
```diff
@@ -10,18 +10,42 @@
...
+from sharc.map.topography import Topography
 
 
 class TopologyInputMap(Topology):
     """
     Generates the coordinates of the BSs based on the base station physical cell data input file.
     """
-
-    def __init__(self, param: ParametersImt):
+    # TODO: topography type hint
```
Comment (brunohcfaria) [inline sharc/topology/topology_input_map.py]: Isn't the type hint already set?

File: tests/parameters_test_input_files.ini [hunk 5]
```diff
@@ -1,7 +1,7 @@
 [GENERAL]
 ###########################################################################
 # Number of simulation snapshots
-num_snapshots = 11
+num_snapshots = 1
```
Comment (brunohcfaria) [inline tests/parameters_test_input_files.ini]: Should change it back to 10.

**Review by brunohcfaria — CHANGES_REQUESTED** [review]
Code looks good. Just some minor changes requested. The whole code must be checked against PEP8 standard before merging.

File: tests/parameters_test_input_files.ini (hunk 5 above)
Comment (CalilQ) [reply tests/parameters_test_input_files.ini]: Should it be 10 or 100?

File: sharc/map/topography.py (hunk 1 above)
Comment (CalilQ) [reply sharc/map/topography.py]: Fixed in commit 87f374bf2411e66e111d28b61c74403673412ea1

File: sharc/map/topography.py (hunk 3 above)
Comment (CalilQ) [reply sharc/map/topography.py]: Removed in commit bc9256d3e8a00988b099fd1021bfbf628c6a0e04

File: sharc/map/topography.py (hunk 2 above)
Comment (CalilQ) [reply sharc/map/topography.py]: The comments were changed in commit bc9256d3e8a00988b099fd1021bfbf628c6a0e04. Pertinent code comments might still be necessary, though.

File: sharc/topology/topology_input_map.py (hunk 4 above)
Comment (CalilQ) [reply sharc/topology/topology_input_map.py]: Yes, it is. That `TODO` is unnecessary and was removed in commit 757d4098ee746558126086ba8b184bdcdf9e6b7a

File: tests/parameters_test_input_files.ini (hunk 5 above)
Comment (CalilQ) [reply tests/parameters_test_input_files.ini]: Changed to 10 by commit 04d6f5de34a7d9a8471c6ed1ce20bae29a3af38d

**Review by brunohcfaria — APPROVED** [review]

**Review by rdvieira — COMMENTED** [review]
Not clear: 
1) what the meaning/value of this parameter utm_zone = 23K; Can I change it randomly?

2) What is the ue distribution function implemented? Where I configured it? There are two other parameters controlling the ue distribution (ue_distribution_distance;ue_distribution_azimuth) how they are correlated with this new implementation?

Enhancements:

1) Move topology parameter to the new parameter_general file.
2) Create a new file with "vale" parameters.

**Comment by brunohcfaria** [comment]
Not clear:

what the meaning/value of this parameter utm_zone = 23K; Can I change it randomly?
>> No. Those are the UTM Zones for UTM projection system.

What is the ue distribution function implemented? Where I configured it? There are two other parameters controlling the ue distribution (ue_distribution_distance;ue_distribution_azimuth) how they are correlated with this new implementation?

>> UEs are distributed with those functions, but now they're dopped inside the polygon 

Enhancements:

Move topology parameter to the new parameter_general file.
Create a new file with "vale" parameters
>> @rdvieira Please add those enhancements to the issue tracker and label it as Enhancements

**Pull request merged**